[open-formulieren/open-forms#3061] Feature: Select Field (#418)
open-formulieren/open-forms-sdk: src/components/forms/SelectField/SelectField.stories.mdx

Reviewer: Please add some proze to this documentation as well - explain when to use the `SelectField`, give examples on how to use the `AsyncSelectField` etc.

import {Canvas, Meta, Story} from '@storybook/addon-docs';
import {expect} from '@storybook/jest';
import {userEvent, within} from '@storybook/testing-library';

> import {FormikDecorator} from '../../../story-utils/decorators';

Reviewer: ```suggestion
import {FormikDecorator} from 'story-utils/decorators';
```

import AsyncSelectField from './AsyncSelectField';

> import {SelectField} from './SelectField';

Reviewer: ```suggestion
import SelectField from './SelectField';
```

import {userEvent, within} from '@storybook/testing-library';

import {FormikDecorator} from 'story-utils/decorators';

import AsyncSelectField from './AsyncSelectField';
import SelectField from './SelectField';

export const Template = args => <SelectField {...args} />;
export const DynamicTemplate = args => <AsyncSelectField {...args} />;
const retrieveOptions = async () => {
  await new Promise(resolve => setTimeout(resolve, 1000));
  return [
    {value: 'option1', label: 'Option 1'},
    {value: 'option2', label: 'Option 2'},
    {value: 'option3', label: 'Option 3'},
  ];
};
const test = async canvasElement => {
  const canvas = within(canvasElement);
  await expect(canvas.getByRole('combobox')).toBeVisible();
  await expect(canvas.getByText('Option')).toBeVisible();
  await expect(canvas.getByText('This is a custom description for the option field')).toBeVisible();
  const label = canvas.getByText('Option');
  userEvent.click(label);
  await expect(canvas.getByRole('combobox')).toHaveFocus();
};

<Meta
  title="Pure React Components / Forms / SelectField"
  component={SelectField}
  parameters={{
    initialValues: {
      test: 'test',
    },
  }}
/>

# Select Field

> We should make use of the SelectField component whenever we have `options` that are statically
> available, and we don't need to retrieve them from an API endpoint. Whenever we **do** need to
> retrieve the options from an API endpoint, we should instead use the AsyncSelectField component.

Reviewer: ```suggestion
Use the `SelectField` component whenever you have `options` that are known upfront. If your
options are retrieved from some external data source (like an API endpoint) instead, use the 
`AsyncSelectField` variant instead.
```

- address the reader
- imperative form makes it easier to extract what to do when, as a reader

## SelectField

<Canvas>
  <Story
    name="Static"
    decorators={[FormikDecorator]}
    play={async ({canvasElement}) => await test(canvasElement)}
    args={{
      name: 'option',
      id: 'option',
      label: 'Option',
      description: 'This is a custom description for the option field',
      disabled: false,
      invalid: false,
      isRequired: true,
      options: [
        {value: 'test', label: 'Test'},
        {value: 'test2', label: 'Test 2'},
      ],
      getOptionValue: option => option.value,
    }}
  >
    {Template.bind({})}
  </Story>
</Canvas>

## AsyncSelectField

> When using this component, we provide a `optionsRetriever` callback that will be used to retrieve
> the options. This callback should return a promise that resolves to an array of options.

Reviewer: ```suggestion
When using this component, provide an `optionsRetriever` callback that will be used to retrieve
the options. This callback should be an async function returning an array of options.
```

async/await is the easier-to-understand abstraction, so while not incorrect, it's advisable to steer people towards using the modern syntax instead of letting them think about promises.
